src/dex/jarvis-v6/utils.ts

```diff
@@ -0,0 +1,130 @@
+import { ethers } from 'ethers';
+import { IDexHelper } from '../../dex-helper';
+import { Address, Token } from '../../types';
+import { bigIntify } from '../nerve/utils';
+import { JarvisSwapFunctions, PoolConfig, PoolState } from './types';
+import SynthereumPriceFeedABI from '../../abi/jarvis/SynthereumPriceFeed.json';
+import { Interface } from '@ethersproject/abi';
+import { BI_POWS } from '../../bigint-constants';
+
+export const THIRTY_MINUTES = 60 * 30;
+
+export async function getOnChainState(
+  dexHelper: IDexHelper,
+  priceFeedAddress: Address,
+  poolConfigs: PoolConfig[],
+  poolInterface: Interface,
+  blockNumber: number | 'latest',
+): Promise<PoolState[]> {
+  const pricesFeedResults = await _getPricesFeedValue(
+    dexHelper,
+    poolConfigs,
+    priceFeedAddress,
+    blockNumber,
+  );
+
+  const multiContract = dexHelper.multiContract;
+  const PoolCallData = poolConfigs
+    .map(pool => [
+      {
+        target: pool.address,
+        callData: poolInterface.encodeFunctionData('feePercentage', []),
+      },
+    ])
+    .flat();
+  const poolDataCalled = await multiContract.methods
+    .aggregate(PoolCallData)
+    .call({}, blockNumber);
+
+  let i = 0;
+  return poolConfigs.map((pool, index) => {
+    return {
+      priceFeed: {
+        usdcPrice: bigIntify(pricesFeedResults[index]),
+      },
+      pool: {
+        feesPercentage: bigIntify(
+          poolInterface.decodeFunctionResult(
+            'feePercentage',
+            poolDataCalled.returnData[i++],
+          )[0],
+        ),
+      },
+    };
+  });
+}
+
+async function _getPricesFeedValue(
+  dexHelper: IDexHelper,
+  poolConfigs: PoolConfig[],
+  priceFeedAddress: Address,
+  blockNumber: number | 'latest',
+) {
+  const pairList = poolConfigs.map(pool =>
+    ethers.utils.formatBytes32String(pool.priceFeedPair),
+  );
+
+  const priceFeedContract = new dexHelper.web3Provider.eth.Contract(
+    SynthereumPriceFeedABI as any,
+    priceFeedAddress,
+  );
+
+  const pricesFeedValues = await priceFeedContract.methods
+    .getLatestPrices(pairList)
+    .call({}, blockNumber);
+
+  return pricesFeedValues;
+}
+
+export function getJarvisPoolFromTokens(
```

Comment:
any way we can discover pools dynamically through onchain registry or similar?

Reply:
Agree to move on a dynamic way to get pools info, but for the next version of the integration

Reply:
For posterity we can achieve this with:
- https://gitlab.com/jarvis-network/apps/exchange/mono-repo/-/blob/dev/libs/contracts/contracts/core/registries/Registry.sol
- https://gitlab.com/jarvis-network/apps/exchange/mono-repo/-/blob/dev/libs/contracts/contracts/core/registries/PoolRegistry.sol

as suggested by @Verisana